### DISCUSSION TITLE
Tree Shaking?

### DISCUSSION THREAD

**Discussion by literalpie** [storybookjs/storybook#16441] [General] [answered]
I'm curious what the current expected behavior of tree shaking in a production build of storybook is. I see that tree shaking is briefly mentioned [here](https://github.com/storybookjs/storybook/issues/9290), but that issue hasn't had much movement recently. [this comment](https://github.com/storybookjs/storybook/issues/9282#issuecomment-815286710) says that tree shaking has been added, but I don't see any other evidence of it.

When I look at the bundle produced by a storybook build, I see a lot of code that I know is never referenced from any of my stories. Even when I change the configuration to only load _one_ story, the bundle size is still very large and includes code from components that are not used. The largest file output by webpack is 7.2 MiB (no gzipping). 

By comparison, the main entry point of our Angular dev app is 1.3MB It's not a true/fair comparison because we have more demos in storybook than the dev app, but I would still expect the entry point to be similar, especially with storyStoreV7 enabled. Also, Storybook is still bigger even when I am only loading one story.

I know that performance has been a big focus, and I've appreciated the recent changes. 
* I'm curious if any work on tree shaking has been done recently, or is in progress?
* Is tree shaking expected to be working today, and I should do more debugging on why it's not working in my project? 
* Maybe there is a difference in behavior between React and Angular projects?

**Comment by literalpie** [comment, marked as answer]
Sorry, I was mistaken. Tree shaking does seem to work. Using the source-map-explorer, I confirmed that certain modules of the Angular CDK library are not included when I remove all references to them from my code.

I think the remaining issue causing the large main bundle size is that all dependencies are included in the main bundle, even if they are only used in certain stories. Ideally, they would be code-split along with the story code. Since the code-splitting feature is so new, I don't have any expectations that it will work optimally right now. I'm hopeful that things will get better in coming releases.

**Reply by shilman** [reply]
Thanks for the investigation @literalpie! The sweeping changes in 6.4 are foundational and we intend to continue optimizing in 6.5 and beyond!

**Comment by herzaso** [comment]
Maybe I'm doing something wrong but I don't think that it's actually working.

Here's a build of a library with a single line `export const a = 1;`:
```
vite v5.0.12 building for production...
✓ 4 modules transformed.
dist/index.js  0.03 kB │ gzip: 0.05 kB
dist/index.js  0.31 kB │ gzip: 0.23 kB
✓ built in 34ms
```

and here's the same library exporting just `{ DocsContext } from '@storybook/blocks';`:
```
vite v5.0.12 building for production...
node_modules/telejson/dist/index.mjs (1413:15) Use of eval in "node_modules/telejson/dist/index.mjs" is strongly discouraged as it poses security risks and may cause issues with minification.
node_modules/telejson/dist/index.mjs (1416:18) Use of eval in "node_modules/telejson/dist/index.mjs" is strongly discouraged as it poses security risks and may cause issues with minification.
✓ 579 modules transformed.
dist/index.js                                    0.09 kB │ gzip:   0.10 kB
dist/WithTooltip-V3YHNWJZ-6OpWCEwN.js            0.16 kB │ gzip:   0.13 kB
dist/syntaxhighlighter-V7JZZA35-2Na3w5AI.js      0.17 kB │ gzip:   0.15 kB
dist/Color-6VNJS4EI-1ZEXpnyR.js                 40.48 kB │ gzip:  12.84 kB
dist/formatter-SWP5E3XI-RegLD4fh.js            849.92 kB │ gzip: 205.05 kB
dist/index-nKSQ2Ucn.js                       1,259.71 kB │ gzip: 325.50 kB
No name was provided for external module "react" in "output.globals" – guessing "React5__default".
dist/index.js  1,563.78 kB │ gzip: 492.93 kB
✓ built in 2.47s
```

this is with Storybook 7.6.5, react 18.2.0, vite 5.0.8 and typescript 5.2.2

**Reply by herzaso** [reply]
repo can be found [here](https://github.com/herzaso/storybook-blocks)
